### PR TITLE
initial macOS support

### PR DESCRIPTION
**Pull request by aconchillo**
Initial macOS support with as minimum changes as I could.

* README.md: The way to change from guile-3.0 to guile-2.2 depends on the system (use `update-alternatives` in Debian, or set the right PATH in macOS).

* dev/prep-for-macos-build: install macOS dependencies through Homebrew.

* added github actions for macos.

### CODE REVIEW

**Comment by aconchillo** [comment]
@rlbdv Two tests are failing in `macos-guile-3`

```
ERROR: test/example-nbody - too few tests run (expected 1, got 0)
ERROR: test/example-nbody - exited with status 2
```

Almost there. I'll try to figure out what's going on with guile 2.

File: README.md [hunk 1]
```diff
@@ -91,10 +117,9 @@ Hopefully the tests will pass.  If not, please report them to the
 fully supported, so depending on the host, something like `make -j5
 check` can be much faster.
 
-If you have more than one version of Guile installed, you may be able
-to select a particular version at configuration time like this:
-
-    $ ./configure GUILE_EFFECTIVE_VERSION=3.0
+If you have more than one version of Guile installed, we suggest to select it
+using your system tools (e.g. `update-alternatives` in Debian based systems or
+using the appropiate PATH in macOS).
```
Comment (vsimon) [inline README.md:122]: nit: appropriate

Comment (aconchillo) [reply README.md]: done.

**Comment by aconchillo** [comment]
guile 2.2 fixed. now these tests are failing in both:

```
ERROR: test/example-nbody - too few tests run (expected 1, got 0)
ERROR: test/example-nbody - exited with status 2
```

**Comment by rlbdv** [comment]
Thanks - could you try this branch?
  https://github.com/rlbdv/lokke/tree/tmp-binary-executables
If that works, then I might like to merge it and we can rebase the rest of your improvements on top.

**Comment by rlbdv** [comment]
Oh, and given the autotools-related changes you might or might not need to either work from a clean tree, or run some or all of the initial setup commands, i.e. if all else fails:
```
autoreconf -fi
./configure
make clean
make  # or make -j5 or whatever
```

**Comment by aconchillo** [comment]
> Thanks - could you try this branch?
> https://github.com/rlbdv/lokke/tree/tmp-binary-executables
> If that works, then I might like to merge it and we can rebase the rest of your improvements on top.

Much better, but there are still issues with the `clojure` related tests that use `./lokke -0` instead of `./guile`. The problem is the same I found yesterday, all arguments in the first line are treated as a list of single arguments. Added a `(pk args)` to `main.scm` and it shows:

```
;;; (("./lokke" "-0" ";;" "-*-clojure-*-" "./test/clojure-basics"))
```

**Comment by rlbdv** [comment]
Hmm, so I just tested on FreeBSD 12.2, which I'd expect to have similar behaviors on this front, and `test/clojure-collection` worked fine, as did `gmake check` on the branch.  I wondered if there might be a chance you had lingering changes with respect to `tmp-binary-executables`.  (If you're not sure, suppose you could try a fresh clone or reset of that branch.)

On the other hand, maybe it is something macos specific.

On the positive side, with some minor changes, things seem to work fine on FreeBSD now.  I'll queue those up for my next update.

**Comment by aconchillo** [comment]
> Hmm, so I just tested on FreeBSD 12.2, which I'd expect to have similar behaviors on this front, and `test/clojure-collection` worked fine, as did `gmake check` on the branch. I wondered if there might be a chance you had lingering changes with respect to `tmp-binary-executables`. (If you're not sure, suppose you could try a fresh clone or reset of that branch.)
> 
> On the other hand, maybe it is something macos specific.
> 
> On the positive side, with some minor changes, things seem to work fine on FreeBSD now. I'll queue those up for my next update.

I cloned your repo fresh and same problem:

```
❯ cat test/clojure-comparison.log
Backtrace:
In ice-9/boot-9.scm:
  1736:10  5 (with-exception-handler _ _ #:unwind? _ # _)
In unknown file:
           4 (apply-smob/0 #<thunk 10a23b5a0>)
In lokke/main.scm:
   304:38  3 (lokke-main _)
   262:25  2 (run-script _ _)
In ice-9/ports.scm:
   440:11  1 (call-with-input-file ";;" #<procedure 10adb5e40 at lo?> ?)
In unknown file:
           0 (open-file ";;" "r" #:encoding #f #:guess-encoding #f)

ERROR: In procedure open-file:
In procedure open-file: No such file or directory: ";;"
```

**Comment by aconchillo** [comment]
> On the other hand, maybe it is something macos specific.

I think that might be the case. macOS splits all arguments and that might be different than FreeBSD.

I still think that keeping the mode line out of the first line would be a good idea, for the same reason you had to do:

```
#!./guile
!# ;; -*-scheme-*-
```

**Comment by aconchillo** [comment]
One last think, Emacs also says you should not use first line

```
In shell scripts, the first line is used to identify the script interpreter, so you cannot put any local variables there. To accommodate this, Emacs looks for local variable specifications in the second line if the first line specifies an interpreter.
```

From https://www.gnu.org/software/emacs/manual/html_node/emacs/Specifying-File-Variables.html

**Comment by rlbdv** [comment]
I just updated main with I think full support for FreeBSD, and rebased `tmp-binary-executables` on that.  Though I suspect it might not improve the macos issue yet.  And while I'm not opposed to moving the mode lines if we need to, I'd first like to make sure I understand the underlying issue because while it's true in general that you shouldn't put it there, I believe what's actually allowed should depend on the #! interpreter, and we're now the (sole) interpreter.

**Comment by rlbdv** [comment]
Oh, and originally (when working on `(lokke main)` a good while back) I'd meant for this to work (specifically wrt `-0`, so now that I have a minute, I'll go remember exactly what I'd intended in there, and I might have just made a mistake.

**Comment by rlbdv** [comment]
OK, had time to investigate, and agree that best solution is the one you've suggested, i.e. forbid arguments other than `-0` on the `#!` line.  I went ahead and updated just that bit and lokke.1, and added some checking in main.  I also pushed the "binary executable" work, so if you'd like to rebase on top of main, we can pursue the tests.

Thanks again for the help.

**Comment by aconchillo** [comment]
> OK, had time to investigate, and agree that best solution is the one you've suggested, i.e. forbid arguments other than `-0` on the `#!` line. I went ahead and updated just that bit and lokke.1, and added some checking in main. I also pushed the "binary executable" work, so if you'd like to rebase on top of main, we can pursue the tests.
> 
> Thanks again for the help.

Thank you! I think it all works now (actually I'm now having issues compiling with Guile 3.0.5 because of some deprecations).

So my PR ended up in just basically README.md updates and other minus things. I commented out `is_latin1_plusminus` since it's unused.

**Comment by rlbdv** [comment]
Aleix Conchillo Flaqué <notifications@github.com> writes:

> So my PR ended up in just basically README.md updates and other minus
> things. I commented out `is_latin1_plusminus` since it's unused.

Oh, right, I think I've removed that in main?

**Comment by aconchillo** [comment]
> Aleix Conchillo Flaqué <notifications@github.com> writes:
> So my PR ended up in just basically README.md updates and other minus things. I commented out `is_latin1_plusminus` since it's unused.
> Oh, right, I think I've removed that in main?

You are right, that was a merge conflict I solved wrong. Removed and PR updated.

**Comment by rlbdv** [comment]
OK, I think I've incorporated all of this into main (and then reworked the broader workflow a bit).  I also removed the CFLAGS and and CFLAGS settings since I think we don't actually need them (I think configure determines those via pkgconf).  In any case, thanks again for the help.

**Comment by aconchillo** [comment]
Cool, thank you! @rlbdv let me know if you are interested in the README update or the github actions for macos which were available in this PR. Happy to help.

**Comment by rlbdv** [comment]
Aleix Conchillo Flaqué <notifications@github.com> writes:

> Cool, thank you! @rlbdv let me know if you are interested in the
> README update or the github actions for macos which were available in
> this PR. Happy to help.

Oh, I think I included them?  I just rearranged things a bit after
incorporating your work.  i.e. macos is included in the main.yaml
workflow, and your code was folded in to that and the new
ci/bin/prep-and-test, etc.  Your documentation changes ended up mostly
in the new INSTALL file.  There should be some additional information in
the relevant commit messages.

But please let me know if what we ended up with doesn't seem suitable.

Thanks again

**Comment by aconchillo** [comment]
Duh, I see it now. All good!

**Pull request closed without merging**